dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java

```diff
@@ -0,0 +1,317 @@
+        String avvisSemantic = ctx.avvisSemantic().content;
+        if (!mo.isEmpty() && !mo.equals(Tag.Vulkan.ACQUIRE) && !avvis.isEmpty() && !avvis.equals(Tag.Vulkan.VISIBLE)) {
+            throw new ParsingException("Loads must be acquire or visible");
+        }
+        Load load = EventFactory.newLoadWithMo(register, location, mo);
+        tagChecker(load, atomic, mo, avvis, scope, avvisSemantic);
+        load.addTags(storageClass, classSemantic);
+        return programBuilder.addChild(mainThread, load);
+    }
+
+    @Override
+    public Object visitRmwConstant(LitmusVulkanParser.RmwConstantContext ctx) {
+        Register register = (Register) ctx.register().accept(this);
+        MemoryObject location = programBuilder.getOrNewMemoryObject(ctx.location().getText());
+        IConst constant = (IConst) ctx.constant().accept(this);
+        Boolean atomic = ctx.atomic().isAtomic;
+        String mo = ctx.mo().content;
+        String avvis = ctx.avvis().content;
+        String scope = ctx.scope().content;
+        String storageClass = ctx.storageClass().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        String avvisSemantic = ctx.avvisSemantic().content;
+        if (!mo.isEmpty() && !mo.equals(Tag.Vulkan.ACQ_REL)) {
+            throw new ParsingException("RMW must be acq_rel");
+        }
+        VulkanRMW rmw = EventFactory.Vulkan.newRMW(location, register, constant, mo, scope);
+        tagChecker(rmw, atomic, mo, avvis, scope, avvisSemantic);
+        rmw.addTags(storageClass, classSemantic, Tag.Vulkan.ATOM);
+        return programBuilder.addChild(mainThread, rmw);
+    }
+
+    @Override
+    public Object visitMemoryBarrier(LitmusVulkanParser.MemoryBarrierContext ctx) {
+        String mo = ctx.mo().content;
+        String avvis = ctx.avvis().content;
+        String scope = ctx.scope().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        if (!mo.equals(Tag.Vulkan.ACQUIRE) && !mo.equals(Tag.Vulkan.RELEASE) && !mo.equals(Tag.Vulkan.ACQ_REL)
+                && !avvis.equals(Tag.Vulkan.AVAILABLE) && !avvis.equals(Tag.Vulkan.VISIBLE)) {
+            throw new ParsingException("Fences must be acquire, release, acq_rel, available or visible");
+        }
+        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
+        tagChecker(fence, false, mo, avvis, scope, "");
+        fence.addTags(classSemantic);
+        return programBuilder.addChild(mainThread, fence);
+    }
+
+    @Override
+    public Object visitControlBarrier(LitmusVulkanParser.ControlBarrierContext ctx) {
+        String scope = ctx.scope().content;
+        String mo = ctx.mo().content;
+        String avvis = ctx.avvis().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        Expression fenceId = (Expression) ctx.barID().accept(this);
+        Fence fence = EventFactory.newFenceWithId(ctx.getText().toLowerCase(), fenceId);
```

Comment:
This results in some weird name like `cbar.acq_rel.wg.semsc00` where the last `0` is actually the id. I think you should simply use `cbar` as the name

Reply:
Updated